src/librustc_typeck/astconv.rs

```diff
@@ -1277,8 +1277,18 @@ impl<'o, 'gcx: 'tcx, 'tcx> dyn AstConv<'gcx, 'tcx>+'o {
         // Find the type of the associated item, and the trait where the associated
         // item is declared.
         let bound = match (&ty.sty, ty_path_def) {
+            (&ty::Adt(adt_def, _), Def::TyAlias(_)) if adt_def.is_enum() => {
+                // Variant of enum resolved as type alias
+                let variant_def = adt_def.variants.iter().find(|vd| vd.name == assoc_name.name);
+                return if let Some(variant_def) = variant_def {
+                    let def = Def::Variant(variant_def.did);
+                    (ty, def)
+                } else {
+                    (tcx.types.err, Def::Err)
```

Comment:
This is not correct, even if the type is an enum, it still can have non-variant associated types.
We need to fallback from variants to what `associated_path_def_to_ty` did previously.

Reply:
Ah, good point.